Sources/Plasma/FeatureLib/pfDXPipeline/plDXPipeline.cpp

```diff
@@ -6410,20 +6414,25 @@ void    plDXPipeline::IHandleTextureMode(plLayerInterface* layer)
             // If we want to ignore vertex alpha, select arg1
             // If we want to ignore texture alpha, select arg2
             // Otherwise (and normally) multiply the two.
-            fD3DDevice->SetTextureStageState( 0, D3DTSS_ALPHAOP,
-                fLayerState[0].fBlendFlags & hsGMatState::kBlendNoVtxAlpha
-                    ? D3DTOP_SELECTARG1
-                    :   fLayerState[0].fBlendFlags & hsGMatState::kBlendNoTexAlpha
-                        ? D3DTOP_SELECTARG2
-                        : D3DTOP_MODULATE );
-            fD3DDevice->SetTextureStageState( 0, D3DTSS_ALPHAARG1, 
-                fLayerState[0].fBlendFlags & hsGMatState::kBlendInvertAlpha 
-                    ? D3DTA_TEXTURE | D3DTA_COMPLEMENT 
-                    : D3DTA_TEXTURE);
-            fD3DDevice->SetTextureStageState( 0, D3DTSS_ALPHAARG2, D3DTA_DIFFUSE | 
-                        ( fLayerState[0].fBlendFlags & hsGMatState::kBlendInvertVtxAlpha 
-                            ? D3DTA_COMPLEMENT 
-                            : 0 ) ); 
+            if (fLayerState[0].fBlendFlags & hsGMatState::kBlendNoVtxAlpha) {
+                fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAOP, D3DTOP_SELECTARG1);
+            } else {
+                if (fLayerState[0].fBlendFlags & hsGMatState::kBlendNoTexAlpha) {
+                    fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAOP, D3DTOP_SELECTARG2);
+                } else {
+                    fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAOP, D3DTOP_MODULATE);
+                }
+            }
+            if (fLayerState[0].fBlendFlags & hsGMatState::kBlendInvertAlpha) {
+                fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAARG1, D3DTA_TEXTURE | D3DTA_COMPLEMENT);
+            } else {
+                fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAARG1, D3DTA_TEXTURE);
+            }
+            if (fLayerState[0].fBlendFlags & hsGMatState::kBlendInvertVtxAlpha) {
+                fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAARG2, D3DTA_DIFFUSE | D3DTA_COMPLEMENT);
+            } else {
+                fD3DDevice->SetTextureStageState(0, D3DTSS_ALPHAARG2, D3DTA_DIFFUSE | 0);
+            }
```

Comment:
Going to weigh in on all the ternary operator changes here:
- I think there is a case that nested ternary operators are hard to read and should be split up.
- I think single level ternary operators are ok - and I think that removing them in favor of if statements is a regression.
- I'm still thinking about the if statements. I'm not sure it's totally clear either and I had to check it several times to make sure it was ok. Nested if statements are their own readability problem. It's possible that maybe the second level of the if statement should be a ternary operator.

Reply:
I'll dig in a bit more on why I think the ternary operator changes are a regression here - and why the original author (who I'm guessing is Cyan) probably used them.

Ternary operators and if/then/else are _not_ implicitly the same thing. A ternary operator guarantees an outcome will be reached, while if/then/else does not.

So when I read:
```
fD3DDevice->SetTextureStageState( 0, D3DTSS_ALPHAARG1, 
                fLayerState[0].fBlendFlags & hsGMatState::kBlendInvertAlpha 
                    ? D3DTA_TEXTURE | D3DTA_COMPLEMENT 
                    : D3DTA_TEXTURE);
```

I _know_ SetTextureStageState state will be called with something. It's guaranteed that SetTextureStageState will still be called. It may take me a bit of grok'ing to figure out what the outcome will be, but I know there will be an outcome.

If/then/else loses that guarantee which is probably why it was not originally used. You have the overhead of grok'ing if an if/then/else will always result in an outcome regardless of the input, which is more complicated and introduces higher risk of error. If/then/else also repeats the same statement over and over - which increases risk that if the programmer needs to make changes they might forget to update one of the if/then branches.

I think there are more efficient ways of breaking up the ternary operators. A ternary operator could be split over several lines if it's nested. And as I said in the original comment - I think un-nested ternary operators should not be removed.